src/unix.rs

```diff
@@ -88,51 +81,41 @@ impl Client {
         Ok(Client::from_fds(pipes[0], pipes[1]))
     }
 
-    pub unsafe fn open(s: &str) -> Result<Client, ErrFromEnv> {
-        match (Self::from_fifo(s), Self::from_pipe(s)) {
-            (Some(Ok(c)), _) | (_, Some(Ok(c))) => Ok(c),
-            (Some(Err(e)), _) | (_, Some(Err(e))) => Err(e),
-            (None, None) => Err(ErrFromEnv::ParseEnvVar),
-        }
+    pub unsafe fn open(s: &str) -> io::Result<Client> {
+        Ok(Self::from_fifo(s)?.unwrap_or(Self::from_pipe(s)?))
     }
 
     /// `--jobserver-auth=fifo:PATH`
-    fn from_fifo(s: &str) -> Option<Result<Client, ErrFromEnv>> {
+    fn from_fifo(s: &str) -> io::Result<Option<Client>> {
         let mut parts = s.splitn(2, ':');
         if parts.next().unwrap() != "fifo" {
-            return None;
+            return Ok(None);
         }
-        let path = match parts.next() {
-            Some(p) => Path::new(p),
-            None => return Some(Err(ErrFromEnv::ParseEnvVar)),
-        };
-        let file = match OpenOptions::new().read(true).write(true).open(path) {
-            Ok(f) => f,
-            Err(e) => return Some(Err(ErrFromEnv::OpenFile(e.to_string()))),
-        };
-        Some(Ok(Client::Fifo {
+        let path = Path::new(parts.next().ok_or(io::Error::new(
```

Comment:
Use ok_or_else to only create io::Error on failure, since io::Error::new does heap allocation internally.